### PR TITLE
[SPARK-36291][SQL] Refactor second set of 20 in QueryExecutionErrors to use error classes

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?

Refactor some exceptions in QueryExecutionErrors to use error classes.

```
inputTypeUnsupportedError
invalidFractionOfSecondError
overflowInSumOfDecimalError
overflowInIntegralDivideError
mapSizeExceedArraySizeWhenZipMapError
copyNullFieldNotAllowedError
literalTypeUnsupportedError
noDefaultForDataTypeError
doGenCodeOfAliasShouldNotBeCalledError
orderedOperationUnsupportedByDataTypeError
regexGroupIndexLessThanZeroError
regexGroupIndexExceedGroupCountError
invalidUrlError
dataTypeOperationUnsupportedError
mergeUnsupportedByWindowFunctionError
dataTypeUnexpectedError
typeUnsupportedError
negativeValueUnexpectedError
addNewFunctionMismatchedWithFunctionError
cannotGenerateCodeForUncomparableTypeError
```
### Why are the changes needed?

There are currently ~350 exceptions in this file; so this PR only focuses on the second set of 20.

### Does this PR introduce _any_ user-facing change?
No
### How was this patch tested?
Existed UT Testcase

### CODE REVIEW

**Comment by dgd-contributor** [comment]
> There is no suitable SQLSTATE for all error class added newly ?

No I haven't working on adding sqlstate, I will add them shortly

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -89,13 +122,31 @@
...
+  "SHOULD_NOT_CALL_ALIAS_DO_GEN_CODE" : {
+    "message" : [ "Alias.doGenCode should not be called." ]
+  },
+  "SUM_OF_DECIMAL_OVERFLOW" : {
+    "message" : [ "Overflow in sum of decimals." ]
+  },
   "UNABLE_TO_ACQUIRE_MEMORY" : {
     "message" : [ "Unable to acquire %s bytes of memory, got %s" ]
   },
+  "UNEXPECTED_DATATYPE" : {
```
[the remark this answers is not in the document]
Comment (dgd-contributor) [reply core/src/main/resources/error/error-classes.json:147]: can we merge them like
```
"UNEXPECTED" : {
    "message" : [ "Unexpected %s" ],
    "sqlState" : "22000"
```

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -6,6 +6,10 @@
   "CONCURRENT_QUERY_ERROR" : {
     "message" : [ "Another instance of this query was just started by a concurrent session." ]
   },
+  "COPY_NULL_FIELD" : {
+    "message" : [ "Do not attempt to copy a null field" ],
+    "sqlState" : "39004"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:11]: This only applies to external routine invocations.

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -14,13 +18,21 @@
     "message" : [ "Found duplicate keys '%s'" ],
     "sqlState" : "23000"
   },
+  "EXCEED_ARRAY_SIZE_WHEN_ZIP_MAP" : {
+    "message" : [ "Unsuccessful try to zip maps with %s unique keys due to exceeding the array size limit %s." ],
+    "sqlState" : "22026"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:23]: We can leave this blank for now; I don't see an obvious SQLSTATE for exceeding the array size limit (22026 is only for string data).

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -32,20 +44,36 @@
   "GROUPING_SIZE_LIMIT_EXCEEDED" : {
     "message" : [ "Grouping sets size cannot be greater than %s" ]
   },
+  "GROUP_INDEX_EXCEED_GROUP_COUNT" : {
+    "message" : [ "Regex group count is %s, but the specified group index is %s" ],
+    "sqlState" : "07009"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:49]: I'm not sure if 07009 is right for group index, it seems to apply to descriptor index. Maybe 42000?

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (karenfeng) [reply core/src/main/resources/error/error-classes.json:147]: I think it makes sense to just keep one or the other; maybe `UNEXPECTED_TYPE`?

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -32,20 +44,36 @@
...
+  },
   "IF_PARTITION_NOT_EXISTS_UNSUPPORTED" : {
     "message" : [ "Cannot write, IF NOT EXISTS is not supported for table: %s" ]
   },
   "INCOMPARABLE_PIVOT_COLUMN" : {
     "message" : [ "Invalid pivot column '%s'. Pivot columns must be comparable." ],
     "sqlState" : "42000"
   },
+  "INCOMPARABLE_TYPE" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:62]: I think this should be more specific; eg. `CANNOT_GENERATE_CODE_FOR_UNCOMPARABLE_TYPE`.

File: core/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -32,20 +44,36 @@
...
+    "message" : [ "cannot generate %s code for un-comparable type: %s" ],
+    "sqlState" : "07006"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:64]: 0A000 may be a better fit

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 7]
```diff
@@ -153,90 +153,88 @@ object QueryExecutionErrors {
...
+      messageParameters = Array(frequencyExpression.sql), cause = null)
   }
 
   def addNewFunctionMismatchedWithFunctionError(funcName: String): Throwable = {
-    new IllegalArgumentException(s"$funcName is not matched at addNewFunction")
+    new SparkIllegalArgumentException("NOT_MATCH_FUNCTION_NAME", Array(funcName))
   }
 
-  def cannotGenerateCodeForUncomparableTypeError(
+  def cannotGenerateCodeForIncomparableTypeError(
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:235]: "Uncomparable" is more accurate than "incomparable": "incomparable" (although more commonly used) generally refers to something that is much better than the others, while "uncomparable" refers to something that cannot be compared to others.

File: core/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -57,6 +85,10 @@
   "INVALID_JSON_SCHEMA_MAPTYPE" : {
     "message" : [ "Input schema %s can only contain StringType as a key type for a MapType." ]
   },
+  "INVALID_URL" : {
+    "message" : [ "Find an invalid url string %s" ],
+    "sqlState" : "22000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:90]: 42000 may be a better fit

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -77,6 +109,18 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "NOT_MATCH_FUNCTION_NAME" : {
+    "message" : [ "%s is not matched at addNewFunction" ],
+    "sqlState" : "07000"
+  },
+  "NOT_SUPPORT_ORDERED_OPERATIONS" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:116]: Can we name this `UNSUPPORTED_ORDERED_OPERATIONS`?

File: core/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -77,6 +109,18 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "NOT_MATCH_FUNCTION_NAME" : {
+    "message" : [ "%s is not matched at addNewFunction" ],
+    "sqlState" : "07000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:114]: 42000 may be a better fit

File: core/src/main/resources/error/error-classes.json [hunk 11]
```diff
@@ -77,6 +109,18 @@
...
+  },
+  "NOT_SUPPORT_ORDERED_OPERATIONS" : {
+    "message" : [ "Type %s does not support ordered operations" ],
+    "sqlState" : "07006"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:118]: 0A000 may be a better fit

File: core/src/main/resources/error/error-classes.json [hunk 12]
```diff
@@ -89,13 +133,37 @@
     "message" : [ "The second argument of '%s' function needs to be an integer." ],
     "sqlState" : "22023"
   },
+  "SHOULD_NOT_CALL_ALIAS_DO_GEN_CODE" : {
+    "message" : [ "Alias.doGenCode should not be called." ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:138]: We can leave this blank; this is an internal error.

File: core/src/main/resources/error/error-classes.json [hunk 13]
```diff
@@ -89,13 +133,37 @@
...
+  },
+  "SUM_OF_DECIMAL_OVERFLOW" : {
+    "message" : [ "Overflow in sum of decimals." ],
+    "sqlState" : "22015"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:142]: 22015 is for interval fields; maybe 22003 is a better fit?

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (karenfeng) [reply core/src/main/resources/error/error-classes.json:147]: I also think 42000 may be a better fit.

File: core/src/main/resources/error/error-classes.json [hunk 14]
```diff
@@ -104,6 +172,10 @@
     "message" : [ "The target JDBC server does not support transaction and can only support ALTER TABLE with a single action." ],
     "sqlState" : "0A000"
   },
+  "WINDOW_NOT_SUPPORT_MERGING" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:175]: This doesn't make sense grammatically; can we rephrase it along the lines of "MERGE_UNSUPPORTED_BY_WINDOW_FUNCTIONS"?

File: core/src/main/scala/org/apache/spark/SparkException.scala [hunk 15]
```diff
@@ -224,3 +224,70 @@ class SparkSQLFeatureNotSupportedException(errorClass: String, messageParameters
   override def getErrorClass: String = errorClass
   override def getSqlState: String = SparkThrowableHelper.getSqlState(errorClass)
 }
+
+/**
+ * Unsupported operation exception thrown from Spark with an error class.
+ */
+class SparkUnsupportedOperationException(errorClass: String, messageParameters: Array[String])
```
Comment (karenfeng) [inline core/src/main/scala/org/apache/spark/SparkException.scala:231]: These should be package-private, eg. https://github.com/apache/spark/blob/f78d8394dcf19891141e353ea3b6a76020faf844/core/src/main/scala/org/apache/spark/SparkException.scala#L221.

**Review by karenfeng — COMMENTED** [review]
Thanks for working on this @dgd-contributor! I left some comments.

File: core/src/main/resources/error/error-classes.json [hunk 16]
```diff
@@ -14,13 +21,20 @@
     "message" : [ "Found duplicate keys '%s'" ],
     "sqlState" : "23000"
   },
+  "EXCEED_ARRAY_SIZE_WHEN_ZIP_MAP" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:24]: This doesn't make much sense grammatically. How about `ZIPPED_MAPS_EXCEED_SIZE_LIMIT`?

File: core/src/main/resources/error/error-classes.json [hunk 17]
```diff
@@ -46,6 +68,10 @@
     "message" : [ "Index %s must be between 0 and the length of the ArrayData." ],
     "sqlState" : "22023"
   },
+  "INTEGRAL_DIVIDE_OVERFLOW" : {
+    "message" : [ "Overflow in integral divide." ],
+    "sqlState" : "22015"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:73]: I think `22003` is a better fit here.

File: core/src/main/resources/error/error-classes.json [hunk 18]
```diff
@@ -32,6 +46,14 @@
   "GROUPING_SIZE_LIMIT_EXCEEDED" : {
     "message" : [ "Grouping sets size cannot be greater than %s" ]
   },
+  "GROUP_INDEX_EXCEED_GROUP_COUNT" : {
+    "message" : [ "Regex group count is %s, but the specified group index is %s" ],
+    "sqlState" : "42000"
+  },
+  "GROUP_INDEX_LESS_THAN_ZERO" : {
+    "message" : [ "The specified group index cannot be less than zero" ],
+    "sqlState" : "07009"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:55]: I think 42000 is a better fit here as well.

File: core/src/main/resources/error/error-classes.json [hunk 19]
```diff
@@ -14,13 +21,20 @@
...
+    "message" : [ "Unsuccessful try to zip maps with %s unique keys due to exceeding the array size limit %s." ]
+  },
   "FAILED_RENAME_PATH" : {
     "message" : [ "Failed to rename %s to %s as destination already exists" ],
     "sqlState" : "22023"
   },
   "FAILED_SET_ORIGINAL_PERMISSION_BACK" : {
     "message" : [ "Failed to set original permission %s back to the created path: %s. Exception: %s" ]
   },
+  "FOUND_NEGATIVE_VALUES" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:34]: I think `UNEXPECTED_NEGATIVE_VALUES` is a better descriptor.

File: core/src/main/resources/error/error-classes.json [hunk 20]
```diff
@@ -3,9 +3,16 @@
...
     "sqlState" : "42000"
   },
+  "CANNOT_GENERATE_CODE_FOR_UNCOMPARABLE_TYPE" : {
+    "message" : [ "cannot generate %s code for un-comparable type: %s" ],
+    "sqlState" : "0A000"
+  },
   "CONCURRENT_QUERY" : {
     "message" : [ "Another instance of this query was just started by a concurrent session." ]
   },
+  "COPY_NULL_FIELD" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:13]: I think `COPY_NULL_FIELD_NOT_ALLOWED` is a better descriptor.

File: core/src/main/resources/error/error-classes.json [hunk 21]
```diff
@@ -77,6 +111,14 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "NOT_MATCH_FUNCTION_NAME" : {
+    "message" : [ "%s is not matched at addNewFunction" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:116]: This is an internal error, let's remove the SQLSTATE.

File: core/src/main/resources/error/error-classes.json [hunk 22]
```diff
@@ -77,6 +111,14 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "NOT_MATCH_FUNCTION_NAME" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:114]: This doesn't make much sense grammatically. How about `ADD_NEW_FUNCTION_NAME_MISMATCHED`?

File: core/src/main/resources/error/error-classes.json [hunk 23]
```diff
@@ -89,17 +131,40 @@
...
+  },
+  "UNSUPPORTED_INPUT_TYPE" : {
+    "message" : [ "Unsupported input type %s" ],
+    "sqlState" : "0A000"
+  },
   "UNSUPPORTED_LITERAL_TYPE" : {
     "message" : [ "Unsupported literal type %s %s" ],
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_ORDERED_OPERATIONS" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:164]: This should be more specific, like `ORDERED_OPERATIONS_UNSUPPORTED_BY_DATA_TYPE`

File: core/src/main/resources/error/error-classes.json [hunk 24]
```diff
@@ -89,17 +131,40 @@
...
   },
+  "UNEXPECTED_TYPE" : {
+    "message" : [ "Unexpected type %s" ],
+    "sqlState" : "42000"
+  },
   "UNRECOGNIZED_SQL_TYPE" : {
     "message" : [ "Unrecognized SQL type %s" ],
     "sqlState" : "42000"
   },
+  "UNSUPPORTED_DATATYPE" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:152]: Nit: `DATATYPE` -> `DATA_TYPE`.

**Review by karenfeng — COMMENTED** [review]
Thanks for working on this @dgd-contributor! I left some comments, please tag me when you address them.

**Comment by dgd-contributor** [comment]
@karenfeng Can you re-review, I've resolve conflict and what you suggested

File: core/src/main/resources/error/error-classes.json [hunk 25]
```diff
@@ -60,6 +70,14 @@
   "GROUPING_SIZE_LIMIT_EXCEEDED" : {
     "message" : [ "Grouping sets size cannot be greater than %s" ]
   },
+  "GROUP_INDEX_EXCEED_GROUP_COUNT" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:73]: Nit: EXCEED -> EXCEEDS

**Review by karenfeng — APPROVED** [review]
LGTM overall - I only have one small typo nit. @HyukjinKwon, can you take a pass once that's addressed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 26]
```diff
@@ -170,90 +170,90 @@ object QueryExecutionErrors {
   }
 
   def inputTypeUnsupportedError(dataType: DataType): Throwable = {
-    new IllegalArgumentException(s"Unsupported input type ${dataType.catalogString}")
+    new SparkIllegalArgumentException("UNSUPPORTED_INPUT_TYPE", Array(dataType.catalogString))
   }
 
   def invalidFractionOfSecondError(): DateTimeException = {
-    new SparkDateTimeException(errorClass = "INVALID_FRACTION_OF_SECOND", Array.empty)
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:177]: Why was this changed?

**Comment by HyukjinKwon** [comment]
@dgd-contributor, BTW, I think we shouldn't use one account shared to many people because:
1. it is ambiguous who reviewers communicate w/, e.g.) reviewers asked one thing to fix in a PR but other PRs still have the same mistakes
2. it is unclear who's the main contributor in the change

I sent an email to the address in GitHub profile but I haven't got a reply yet. Would you mind contacting me via @gurwls223@apache.org?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 26 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:177]: sorry, just a clean up

**Comment by HyukjinKwon** [comment]
@dgd-contributor, please contact me or private@spark.apache.org. As I shared in the email, the submissions from the specific shared account will not be accepted for now.

**Comment by dgd-contributor** [comment]
> @dgd-contributor, please contact me or [private@spark.apache.org](mailto:private@spark.apache.org). As I shared in the email, the submissions from the specific shared account will not be accepted for now.

Could  you please check mail from ncthanh.work@gmail.com? Sorry for the inconvenience because the email from your side has been blocked by our mail system.
